### ISSUE TITLE
ISLaSolver.solve_quantifier_free_formula() doesn't honor max_number_smt_instantiations

### ISSUE THREAD

**Issue by michaelmera** [rindPHI/isla#4]
Here is a file reproducing the problem:
```Python
# isla_wrong_max_smt.py

from isla.type_defs import Grammar
from isla.solver import ISLaSolver

GRAMMAR: Grammar = {
    "<start>": ["<point>"],
    "<point>": ["<x> <y> <z>"],
    "<x>": ["<digit>"],
    "<y>": ["<digit>"],
    "<z>": ["<digit>"],
    "<digit>": ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"],
}

solver = ISLaSolver(
    GRAMMAR,
    """
    forall <point> seed in start:
        (<= (+ (+ (^ (str.to.int seed.<x>) 2) (^ (str.to.int seed.<y>) 2)) (^ (str.to.int seed.<z>) 2)) 900)
    """,
    max_number_smt_instantiations=20,
)

if __name__ == "__main__":
    for s in solver.solve():
        print(s)

    print("done")
```

The output is:
```console
$ python isla_wrong_max_smt.py
0 0 0
1 8 8
4 2 4
9 4 6
6 6 9
2 3 2
7 1 1
8 7 5
3 9 7
5 5 3
0
done
```

There are many solutions missing and `solver.solve()` still stops before it honors `max_number_smt_instantiations`.

I tracked the problem down to [this loop](https://github.com/rindPHI/isla/blob/c5284c7031/src/isla/solver.py#L1419).
It adds the negated values to avoid repeating the same solution, but for a formula containing several variables, it will add each negated value independently, rather than a negated conjunction.

**Comment by rindPHI** [comment]
Fixed in v0.10.2 (https://github.com/rindPHI/isla/blob/main/tests/test_solver.py#L827)

**Comment by rindPHI** [comment]
Thanks for the detailed problem analysis!!!
